### PR TITLE
Add test checking VDI.clone, VDI.copy, and VDI.snapshot update cbt_enabled field

### PR DESCRIPTION
**Pull request by minishrink**
Currently tests two forms of VDI.copy:
- copying between existing VDIs: base_vdi -> into_vdi 
- copying into newly created VDI: vdi -> sR (-> new_vdi)

### CODE REVIEW

File: ocaml/xapi/quicktest_cbt.ml [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  (* Helper for test failure due to unexpected error *)
+  let report_failure error test_name test =
+    failed test (Printf.sprintf "%s failed: %s" test_name
+                   (ExnHelper.string_of_exn error)) in
+
+  let test_assert ~test op ~msg =
+    try assert op with assert_failure -> failed test msg in
+
+  (* Overall test suite runs smaller unit tests *)
+  let test_compare ~test l_op r_op ~msg =
```
Comment (lindig) [inline ocaml/xapi/quicktest_cbt.ml]: Minor suggestion, why not use simpler names like `left` and `right`?

File: ocaml/xapi/quicktest_cbt.ml [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    let op = (l_op = r_op) in test_assert ~test op ~msg in
+
+  let name_description = "VDI for CBT quicktest" in
+  let make_vdi_from sR = (* SR has VDI.create as allowed *)
+    VDI.create ~session_id ~rpc:!rpc ~name_label:"qt-cbt" ~name_description ~_type:`user
```
Comment (lindig) [inline ocaml/xapi/quicktest_cbt.ml]: I would suggest to format this one parameter per line:
```
VDI.create ~session_id 
  ~rpc:!rpc
  ~name_description
  ~_type:`user
 ...
```

This is easier to read and easier to change. However, this is a matter of taste and just a suggestion.

File: ocaml/xapi/quicktest_cbt.ml [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+          content_id_str
+          ~msg:"VDI.data_destroy failed to update VDI.content_id";
+        VDI.destroy ~session_id ~rpc:!rpc ~self:newvdi;
+        debug data_destroy_test "Snapshot destroyed.";
+        success data_destroy_test
+      with e -> report_failure e "data_destroy" data_destroy_test in
+
+    (* Check VDI.(copy, clone, and snapshot) all properly update cbt_enabled *)
+    let vdi_clone_copy_snapshot_test ~sR ~vDI =
+      let vdi_duplicate_test = make_test "Testing VDI.clone,copy,snapshot" 4 in
```
Comment (lindig) [inline ocaml/xapi/quicktest_cbt.ml]: I'd write `VDI.{clone,copy,snapshot}`.

File: ocaml/xapi/quicktest_cbt.ml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+        List.iter (fun self -> VDI.destroy ~session_id ~rpc:!rpc ~self)
+          [ vdi_clone ; vdi_snapshot ; vdi_copy_fresh ; vdi_copy_existing ; into_vdi];
+        success vdi_duplicate_test
+      with e -> report_failure e "VDI_copy_clone_snapshot" vdi_duplicate_test in
+
+
+    let run_test_suite ~sR ~vDI =
+      let sr_ops = (SR.get_allowed_operations ~session_id ~rpc:!rpc ~self:sR) in
+      List.iter
```
Comment (lindig) [inline ocaml/xapi/quicktest_cbt.ml]: Just a suggestion:

```
[ (fun () -> enable_disable_cbt_test ~vDI), [ `vdi_enable_cbt ; `vdi_disable_cbt ]
; (fun () -> vdi_clone_copy_snapshot_test ~sR ~vDI), [ `vdi_enable_cbt ; `vdi_create ; `vdi_clone ; `vdi_snapshot ; `vdi_destroy ]
; (fun () -> vdi_data_destroy_test ~vDI), [ `vdi_enable_cbt ; `vdi_data_destroy ; `vdi_snapshot ; `vdi_destroy ]
]
|> List.iter (fun (test, ops) -> ... )
```

This way data flows from top to bottom.

File: ocaml/xapi/quicktest_cbt.ml [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+          ; (fun () -> vdi_data_destroy_test ~vDI), [ `vdi_enable_cbt ; `vdi_data_destroy ; `vdi_snapshot ; `vdi_destroy ]
+          ]
+    ; in
+
+    (* Try running test suite, definitively destroy all VDIs created, regardless of success or errors *)
+    let handle_storage_objects ~sR ~vDI =
+      Xapi_stdext_pervasives.Pervasiveext.finally
+        (fun () -> run_test_suite ~sR ~vDI)
+        (fun () ->
+           List.filter (fun vdi -> (VDI.get_name_label ~session_id ~rpc:!rpc ~self:vdi = "qt-cbt")
```
Comment (lindig) [inline ocaml/xapi/quicktest_cbt.ml]:
```
VDI.get_all ~session_id ~rpc:!rpc
|> List.filter (..)
|> List.iter (..)
```

File: ocaml/xapi/quicktest_cbt.ml [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                                   && (VDI.get_name_description ~session_id ~rpc:!rpc ~self:vdi = name_description)
+                       )  (VDI.get_all ~session_id ~rpc:!rpc)
+           |> List.iter (fun vdi -> VDI.destroy ~session_id ~rpc:!rpc ~self:vdi);
+        ) in
+
+    (* Obtain list of SRs capable of creating VDIs, and run them all through test suite *)
+    let srs_can_create_vdi =
+      List.filter (fun sR -> List.mem `vdi_create (SR.get_allowed_operations ~session_id ~rpc:!rpc ~self:sR)
```
Comment (lindig) [inline ocaml/xapi/quicktest_cbt.ml]: Again, using `|>` could make this simpler.

File: ocaml/xapi/quicktest_cbt.ml (hunk 4 above)
Comment (mseri) [reply ocaml/xapi/quicktest_cbt.ml]: I'd go the next step and do 
```ocaml
[ (fun () -> enable_disable_cbt_test ~vDI),
  [ `vdi_enable_cbt ; `vdi_disable_cbt ]
; (fun () -> vdi_clone_copy_snapshot_test ~sR ~vDI),
  [ `vdi_enable_cbt ; `vdi_create ; `vdi_clone ; `vdi_snapshot ; `vdi_destroy ]
; (fun () -> vdi_data_destroy_test ~vDI),
  [ `vdi_enable_cbt ; `vdi_data_destroy ; `vdi_snapshot ; `vdi_destroy ]
]
|> List.iter (fun (test, ops) -> ... )

**Comment by mseri** [comment]
This seems to need rebasing on top of the current feature/CBT branch

**Comment by minishrink** [comment]
Updated and rebased on top of feature/CBT

**Pull request closed without merging**